esp-hal/src/rmt.rs

```diff
@@ -250,59 +250,235 @@ pub enum Error {
+    #[inline]
+    const unsafe fn new_unchecked(
+        level1: Level,
+        length1: u16,
+        level2: Level,
+        length2: u16,
+    ) -> Self {
+        let mut code = 0;
 
-    /// Logical output level in the first pulse code interval
-    fn level1(&self) -> Level;
+        if matches!(level1, Level::High) {
+            code |= 1 << (LEVEL1_SHIFT + 15);
+        }
+        code |= (length1 as u32) << LEVEL1_SHIFT;
 
-    /// Length of the first pulse code interval (in clock cycles)
-    fn length1(&self) -> u16;
+        if matches!(level2, Level::High) {
+            code |= 1 << (LEVEL2_SHIFT + 15);
+        }
+        code |= (length2 as u32) << LEVEL2_SHIFT;
+
+        Self(code)
+    }
+
+    /// Create a new instance.
+    ///
+    /// If `length1` or `length2` exceed the maximum representable range, they
+    /// will be clamped to `0x7FFF`.
+    #[inline]
+    pub const fn new(level1: Level, length1: u16, level2: Level, length2: u16) -> Self {
+        // Can't use lengthX.min(0x7FFF) since it is not const
+        let length1 = if length1 >= 0x8000 { 0x7FFF } else { length1 };
+        let length2 = if length2 >= 0x8000 { 0x7FFF } else { length2 };
+
+        // SAFETY:
+        // - We just clamped length1 and length2 to the required intervals
+        unsafe { Self::new_unchecked(level1, length1, level2, length2) }
+    }
+
+    /// Create a new instance.
+    ///
+    /// If `length1` or `length2` exceed the maximum representable range, this
+    /// will return `None`.
+    #[inline]
+    pub const fn try_new(level1: Level, length1: u16, level2: Level, length2: u16) -> Option<Self> {
+        if length1 >= 0x8000 || length2 >= 0x8000 {
+            return None;
+        }
+
+        // SAFETY:
+        // - We just checked that length1 and length2 have their MSB cleared.
+        Some(unsafe { Self::new_unchecked(level1, length1, level2, length2) })
+    }
+
+    /// Create a new empty instance.
+    ///
+    /// This corresponds to the all-zero `PulseCode`, i.e. an end marker with
+    /// level `Level::Low`.
+    #[inline]
+    pub const fn empty() -> Self {
+        Self(0)
+    }
+
+    /// Create a new instance that is and end marker with `Level::Low`.
+    ///
+    /// This is equivalent to `PulseCode::empty()` and provided as a more
+    /// semantic alias.
+    // FIXME: Consider adding a variant with `level1`, `length1` and `level2` arguments
+    // which sets `length2 = 0` so that it is still guaranteed to return an end
+    // marker.
+    #[inline]
+    pub const fn end_marker() -> Self {
+        Self(0)
+    }
+
+    /// Set all levels and lengths to 0.
+    ///
+    /// In other words, assigns the value of `PulseCode::empty()` to `self`.
+    #[inline]
+    pub fn reset(&mut self) {
+        self.0 = 0
+    }
+
+    /// Logical output level in the first pulse code interval
+    #[inline]
+    pub const fn level1(&self) -> Level {
+        // Can't use Level::from(bool) since it is non-const
```

Comment:
I'd say let's add `const fn const_from` to Level, and use that.

Reply:
Makes sense :+1: Would you make that `pub`, or `pub(crate)`?

Reply:
could be `#[instability::unstable] pub`, but `pub(crate)` sounds fine